drivers/kw2xrf/kw2xrf_rx.c

```diff
@@ -0,0 +1,189 @@
+#include "debug.h"
+
+/* circular buffer for incoming 802.15.4 packets */
+kw2xrf_packet_t kw2xrf_rx_buffer[MKW2XDRF_RX_BUF_SIZE];
+volatile uint8_t rx_buffer_next;  /* index of next free cell in RX buffer */
+
+/* pointer to the callback low-level function for packet reception */
+typedef struct {
+    enum { MKW2XDRF_CB_TYPE_NONE = 0, MKW2XDRF_CB_TYPE_RAW, MKW2XDRF_CB_TYPE_DATA } type;
+    union {
+        netdev_802154_raw_packet_cb_t raw;
+        netdev_rcv_data_cb_t data;
+    } cb;
+} kw2xrf_callback_t;
+
+/* pointer to the callback low-level function for packet reception */
+static kw2xrf_callback_t recv_func = {0, {NULL}};
+
+int kw2xrf_add_raw_recv_callback(netdev_t *dev, netdev_802154_raw_packet_cb_t recv_cb)
+{
+    /* XXX: first check only for backwards compatibility with transceiver
+     *      (see kw2xrf_init) remove when adapter for transceiver exists */
+    if (dev != &kw2xrf_netdev) {
+        return -ENODEV;
+    }
+
+    if (recv_func.cb.raw != NULL) {
+        return -ENOBUFS;
+    }
+
+    recv_func.type = MKW2XDRF_CB_TYPE_RAW;
+    recv_func.cb.raw = recv_cb;
+
+    return 0;
+}
+
+int kw2xrf_rem_raw_recv_callback(netdev_t *dev, netdev_802154_raw_packet_cb_t recv_cb)
+{
+    /* XXX: first check only for backwards compatibility with transceiver
+     *      (see cc2420_init) remove when adapter for transceiver exists */
+    if (dev != &kw2xrf_netdev) {
+        return -ENODEV;
+    }
+
+    if (recv_func.cb.raw == recv_cb) {
+        recv_func.cb.raw = NULL;
+    }
+
+    return 0;
+}
+
+int kw2xrf_add_data_recv_callback(netdev_t *dev, netdev_rcv_data_cb_t recv_cb)
+{
+    /* XXX: first check only for backwards compatibility with transceiver
+     *      (see kw2xrf_init) remove when adapter for transceiver exists */
+    if (dev != &kw2xrf_netdev) {
+        return -ENODEV;
+    }
+
+    if (recv_func.cb.data != NULL) {
+        return -ENOBUFS;
+    }
+
+    recv_func.type = MKW2XDRF_CB_TYPE_DATA;
+    recv_func.cb.data = recv_cb;
+
+    return 0;
+}
+
+int kw2xrf_rem_data_recv_callback(netdev_t *dev, netdev_rcv_data_cb_t recv_cb)
+{
+    /* XXX: first check only for backwards compatibility with transceiver
+     *      (see kw2xrf_init) remove when adapter for transceiver exists */
+    if (dev != &kw2xrf_netdev) {
+        return -ENODEV;
+    }
+
+
+    if (recv_func.cb.data == recv_cb) {
+        recv_func.cb.data = NULL;
+    }
+
+    return 0;
+}
+
+void kw2xrf_rx_handler(void)
+{
+    uint8_t pkt_len, pkt_lqi;
+    int8_t pkt_rssi;
+    bool crc_ok = true;
+    uint8_t buf[MKW2XDRF_MAX_PKT_LENGTH + 1];
+
+    pkt_len = kw2xrf_read_dreg(MKW2XDM_RX_FRM_LEN);
+
+    /* read PSDU */
+    kw2xrf_read_fifo(buf, pkt_len + 1);
+    pkt_lqi = buf[pkt_len];
+    pkt_rssi = 0;
+
+    /* low-level reception mechanism (for MAC layer, among others) */
+    if ((recv_func.type == MKW2XDRF_CB_TYPE_RAW) && (recv_func.cb.raw != NULL)) {
+        recv_func.cb.raw(NULL, buf, pkt_len - 2, pkt_rssi, pkt_lqi, crc_ok);
```

Comment:
The modem automatically checks the CRC in hardware. An interrupt is only triggered in case of a valid crc. Therefore the crc_ok is always true. But as it´s not obviously I commented this context and deleted the crc_ok variable as it is not really used.